### PR TITLE
Add “Show Current File in Finder” menu item (#5001)

### LINKED ISSUE #5001: add Show in Finder menu command
**What you want IINA to do:**

- Conveniently show the currently playing media's file in the Finder via menu command.
 
- Even more conveniently do so with a keyboard shortcut. 

Of the two, I think the keyboard shortcut is more important, so I'd be happy with that without the menu item. I vote for a default keyboard shortcut of `⌘R`.

**What IINA does currently:**

The only way that I can find to do this is by showing the playlist, right-click the item, and choose `Show in Finder`. It is a pain, and  it is not keyboard-shortcut-able.

One can right-click on the filename in the titlebar, then click on the containing folder in the dropdown that appears, but this is also inconvenient and not keyboard-shortcut-able.

**Why you think this should be added:**

It's a common need. 

IINA has a `File:Delete Current File` command, so it seems odd that the more general ability to do whatever you want with the file via the Finder is missing.

**Examples of other projects that have something similar:**

Many apps have such a menu command, though named differently:

`Show in Finder`: XCode, macOS Dock, BBEdit
`Reveal in Finder`: Sublime, Transmission, VLC
`Open in Finder`:  Jetbrains IDEs
`Open File Location`: Typora

### PR DESCRIPTION
**Pull request by ShlomoCode**
- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [x] This implements/fixes issue #5001.

---

**Description:**
Adds a menu bar item (under `File`) to reveal the current file in Finder.
~~For some reason I was unable to set a Cmd+R shortcut, even though I set it through the interface builder, the shortcut doesn't work, so the PR does not currently include a shortcut.~~
Built and tested with XCode 15.1, on MacOS 14.5.

### CODE REVIEW

File: iina/Base.lproj/KeyBinding.strings
Comment (low-batt) [inline iina/Base.lproj/KeyBinding.strings]: I will try and find time for a proper review tomorrow. As per issue #3976 this must be show, not reveal. I'd prefer just `Show in Finder` as that is the common label for such a menu item. I think it is ok that it differs from the label for deleting. See what the other reviewers say.

**Comment by low-batt** [comment]
@svobs Should be able to explain how to add the shortcut for the menu item.

File: iina/Base.lproj/KeyBinding.strings
Comment (low-batt) [reply iina/Base.lproj/KeyBinding.strings]: The way our integration with Crowdin works whenever you change a file in Base.lproj the corresponding file in en.lproj needs to be updated as well. No other language files should be changed. They will be updated through Crowdin.

File: iina/MainMenuActions.swift
Comment (uiryuu) [inline iina/MainMenuActions.swift:47]: Remove trailing spaces

**Comment by uiryuu** [comment]
I think adding the keybinding in "IINA Default" input config and it can be automatically matched and show up in the menu.

File: iina/Base.lproj/MainMenu.xib
Comment (low-batt) [inline iina/Base.lproj/MainMenu.xib]: Similar to KeyBinding.strings, the MainMenu.strings in en.lproj must be updated with the new text.

File: iina/Base.lproj/KeyBinding.strings
Comment (uiryuu) [reply iina/Base.lproj/KeyBinding.strings]: I'd like "Show current file in Finder". "Show in Finder" is kinda weird to me since there's no subject.

**Review by uiryuu — CHANGES_REQUESTED** [review]
Apart from what @low-batt has indicated, I have nothing to say. We should consider adding this into the default configuration, and we can add after merging this PR. The problem is should we add this to all default configs (mpv, VLC, Movist)?

File: iina/Base.lproj/KeyBinding.strings
Comment (ShlomoCode) [reply iina/Base.lproj/KeyBinding.strings]: I don't see a significant difference in popularity
https://github.com/search?q=%22Reveal+in+Finder%22&type=code 10.2k
https://github.com/search?q=%22Show+in+Finder%22&type=code 10.6k
In VSCode it's "Reveal", in XCode it's "Show". Change to Show?

File: iina/Base.lproj/MainMenu.xib
Comment (ShlomoCode) [reply iina/Base.lproj/MainMenu.xib]: Am I supposed to do it manually or is there some sync command in XCode?

File: iina/Base.lproj/KeyBinding.strings
Comment (uiryuu) [reply iina/Base.lproj/KeyBinding.strings]: We had a discussion in #3976 before and the conclusion was to use "show" instead of "reveal" all across IINA. So please change to show to make it consistent with other parts of IINA.

File: iina/MainMenuActions.swift
Comment (ShlomoCode) [reply iina/MainMenuActions.swift:47]: done

File: iina/Base.lproj/KeyBinding.strings
Comment (low-batt) [reply iina/Base.lproj/KeyBinding.strings]: Go with @uiryuu "Show current file in Finder" although I prefer "Show in Finder" as that is the typical Apple menu item, like in Xcode. The capitalization needs to be "Show Current File in Finder" as per Apple HIG.

The way I update the en.lproj strings file for a XIB is to use [ibtool](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/MaintaingYourOwnStringsFiles/MaintaingYourOwnStringsFiles.html#//apple_ref/doc/uid/10000171i-CH19-SW6) like so:
```text
ibtool iina/Base.lproj/MainMenu.xib --generate-strings-file temp.strings
```

Then I run [meld](https://github.com/yousseb/meld):
```text
meld temp.strings iina/en.lproj/MainMenu.strings
```

And merge in only the string I added.

File: iina/Base.lproj/MainMenu.xib
Comment (low-batt) [reply iina/Base.lproj/MainMenu.xib]: See my latest comment on how I use `ibtool` to do it.

File: iina/Base.lproj/KeyBinding.strings
Comment (ShlomoCode) [reply iina/Base.lproj/KeyBinding.strings]: BTW note that the screenshot OSD use "REVEAL" word

File: iina/MainMenuActions.swift
Comment (low-batt) [inline iina/MainMenuActions.swift:50]: Missing check for `isNetworkResource`. Not sure how `activateFileViewerSelecting` will react if given a URL to a video on a web site. Have a look at the code in `PlaylistViewController` that calls `activateFileViewerSelecting`.

**Comment by low-batt** [comment]
We do need to get the menu shortcut working. The one suggested is ⌘R. I checked [Mac keyboard shortcuts](https://support.apple.com/en-us/102650) and this is how Apple uses that key:
> Command-R: (1) When an alias is selected in the Finder: show the original file for the selected alias. (2) In some apps, such as Calendar or Safari, refresh or reload the page. (3) In Software Update, check for software updates again.

I don't think it is unreasonable to use ⌘R, but I'm not that familiar with IINA's key binding code. Need to hear from @svobs and @lhc70000.

File: iina/Base.lproj/KeyBinding.strings
Comment (ShlomoCode) [reply iina/Base.lproj/KeyBinding.strings]: done

File: iina/MainMenuActions.swift
Comment (ShlomoCode) [reply iina/MainMenuActions.swift:50]: The approach taken in the `PlaylistViewController` is to simply ignore such items, and this is indeed necessary to avoid an error when using the keyboard shortcut. I applied it. However, in our case it would have been better to disable the button in advance (and display it as gray). Is it possible?

Comment (ShlomoCode) [reply iina/MainMenuActions.swift:50]: I noticed that the similar item, "Delete current file", throws an error when trying to delete a network file:
341644475-8f15b23e-6b94-4d5d-bb33-abfef2af616f
We will have to apply the check there as well.

**Review by uiryuu — APPROVED** [review]
I'm okay with the changes. I'll try to abstract the logic of `@IBAction func contextMenuShowInFinder(_ sender: NSMenuItem)` after merging this PR. Not sure whether or not that's worthwhile

**Comment by svobs** [comment]
> @svobs Should be able to explain how to add the shortcut for the menu item.

Add the 2 lines below to the case statement of `handleIINACommand`, in `PlayerWindowController.swift`:

```Swift
internal func handleIINACommand(_ cmd: IINACommand) {
    switch cmd {
    // […]
    case .showCurrentFileInFinder:
      menuActionHandler.menuShowCurrentFileInFinder(.dummy)
    // […]
```

**Comment by svobs** [comment]
Having connectivity problems here...strange... After adding the code I posted above, everything will be in place for the end user to add a row like this one to your Key Bindings configuration:
Key column: `Meta+r`
Action column: `#@iina show-current-file-in-finder`

File: iina/MainMenuActions.swift
Comment (ShlomoCode) [reply iina/MainMenuActions.swift:51]: > In our case it would have been better to disable the button in advance (and display it as gray). Is it possible?

@uiryuu?

**Comment by ShlomoCode** [comment]
I confirm that it is now possible to set a keyboard shortcut in the settings and it works.
CleanShot 2024-06-21 at 14 18 08@2x

One thing I noticed that when I use two screens MacOS opens the Finder on the built-in screen and does not focus on it, meaning the focus remains on the IINA window. When I only use the built-in screen it doesn't happen. But I'm guessing this is a macOS bug and not something we should be dealing with.

File: iina/MainMenuActions.swift
Comment (ShlomoCode) [reply iina/MainMenuActions.swift:50]: > In our case it would have been better to disable the button in advance (and display it as gray). Is it possible?

I implemented this using `validateMenuItem`, like in `PlaylistViewController` and `HistoryWindowController`.

We will have to handle `menuDeleteCurrentFile` in the same way (add an internal `guard` to prevent an error when using the keyboard shortcut, and add it to `validateMenuItem` to disable the menu item). But I didn't know if it was appropriate to include it in the current PR, let me know whether to add it as well.

**Comment by svobs** [comment]
> One thing I noticed that when I use two screens MacOS opens the Finder on the built-in screen and does not focus on it, meaning the focus remains on the IINA window. When I only use the built-in screen it doesn't happen. 

Try changing this setting in your Finder dock icon:
SCR-20240621-kqdk

Don't ask me how this is supposed to work though. In fact, if anyone knows what the `None` option does or what it means when multiple options are checked, please let me know 😅

**Comment by svobs** [comment]
> One thing I noticed that when I use two screens MacOS opens the Finder on the built-in screen and does not focus on it, meaning the focus remains on the IINA window. When I only use the built-in screen it doesn't happen

Try changing this option in your Finder's Dock icon:
SCR-20240621-kvod

Don't ask me how this is supposed to work though. In fact, if anyone can explain what the `None` option is supposed to do, or what it means when multiple options are checked, please let me know. 😅

**Comment by svobs** [comment]
> One thing I noticed that when I use two screens MacOS opens the Finder on the built-in screen and does not focus on it, meaning the focus remains on the IINA window. When I only use the built-in screen it doesn't happen.

I believe it has something to do with the `Assign To` setting in the Dock. Try setting this to different values for Finder and for IINA, respectively.
SCR-20240621-kvod

Some combinations appear to focus on the new Finder window and some do not. For me, using `Desktop on Display 1` (main display) for Finder and `All Desktops` for IINA correctly changed the focus. So I suspect when some combinations don't give focus to the Finder window, that is by design for some reason.

I haven't been able to find documentation for this feature. From playing around, it looks like the `All Desktops` option will open windows for the app in the same screen that the mouse is currently in at the time of open. But I can't figure out what the `None` option is supposed to do. And I have no idea why sometimes multiple options are checked. If anyone has any info, please share :)

File: iina/MainMenuActions.swift
Comment (low-batt) [reply iina/MainMenuActions.swift:50]: Lets treat the problem with deleting as a separate issue. I will enter one for it.

Comment (low-batt) [reply iina/MainMenuActions.swift:50]: I entered #5011 for the problem with `Delete Current File` being enabled while streaming.

**Comment by low-batt** [comment]
The shortcut key still needs to be added to `iina-default-input.conf`, yes? I've not looked to see if that key is available in the other key bindings.

File: iina/MainMenuActions.swift
Comment (low-batt) [inline iina/MainMenuActions.swift]: Could combine these.
```suggestion
    guard let url = player.info.currentURL, !player.info.isNetworkResource else { return }
```

**Comment by svobs** [comment]
> Add the 2 lines below to the case statement of handleIINACommand, in PlayerWindowController.swift

I was in a little bit of a hurry when I wrote that, but I found time to re-evaluate and have determined that adding a case to `handleIINACommand` is no longer needed. There is a newer mechanism introduced by the fix to #4567 which makes it mostly obsolete. Although it shouldn't hurt anything to leave it there, it can be removed.

I filed a PR, #5012, which has more info if curious.

File: iina/MainMenuActions.swift
Comment (ShlomoCode) [reply iina/MainMenuActions.swift]: I prefer the explicit form, but I don't know the Swift conventions. I followed your suggestion.

**Comment by ShlomoCode** [comment]
I'm not sure that Cmd+R fits here, because it's usually used to refresh (in browsers, XCode, etc.)
Might we want it for a streaming refresh? Thoughts?

**Comment by low-batt** [comment]
I too am concerned about the choice of key.

From [Spotlight keyboard shortcuts on Mac](https://support.apple.com/guide/mac-help/spotlight-keyboard-shortcuts-mh26783/mac):
> See a file in an app or the Finder
> Command-R or Command-Double-click

So we have an Apple example of ⌘R being used to bring up `Finder`.

I need to defer to @svobs and @lhc70000 on the choice of key.

**Comment by ShlomoCode** [comment]
> I believe it has something to do with the `Assign To` setting in the Dock. Try setting this to different values for Finder and for IINA, respectively.

I use an external screen, not space (macOS virtual screen). It seems to only work for spaces, it doesn't show up for me for the external screen, only when I create two spaces on the same physical screen.
CleanShot 2024-06-23 at 02 32 52@2x
macOS 14.5

**Comment by saagarjha** [comment]
I do want to note that this is feature already pretty much available because we present a document proxy icon, which makes it easy to locate the file and open a Finder window for it. I’m not opposed to also adding a “Reveal in Finder” menu item but it definitely should not be squatting on shortcut prime real estate like ⌘R by default.

**Comment by svobs** [comment]
> I do want to note that this is feature already pretty much available because we present a document proxy icon, which makes it easy to locate the file and open a Finder window for it.

The document icon is not available in legacy full screen or in music mode. It is lost when either the `titled` window style is removed or the window has `titleVisibility = .hidden`. I have bent over backwards trying to figure out how to add it back in manually as one can do with the other title buttons but can't seem to make it work, and can't find documentation anywhere.

**Comment by svobs** [comment]
Also agree that ⌘R is probably not the best default. I would associate it more with Reload or Refresh or some other command starting with R, particularly since the action is no longer called "Reveal in Finder". But I wouldn't use anything with S or F either as those have much stronger existing associations. Maybe just add another modifier key? ⇧⌘R? Looking at the existing IINA defaults, there doesn't seem to be a strong pattern to follow for which modifier(s) to add.

**Comment by low-batt** [comment]
Apple guidance on custom keyboard shortcuts can be found [here](https://developer.apple.com/design/human-interface-guidelines/keyboards#Custom-keyboard-shortcuts). Don't be confused by the entries in the `Recommended usage`  column that are in the wrong row. It indicates the ⌘ key is preferred:
> Prefer the Command key as the main modifier key in a custom keyboard shortcut.

Focusing on the keys that trigger IINA actions we can see there isn't a consistent pattern but many use ⇧⌘:
keys

So ⇧⌘ and one of the keys not already used seems right.

**Comment by uiryuu** [comment]
Not meant to divert the discussion too much, but there's an alternative is to not set any default keybindings for that and let the user to set whatever they want.

**Comment by low-batt** [comment]
I thought about that, but configuring key bindings is a bit advanced. I would not be surprised if many users looked at the new menu item, saw there was no keyboard shortcut and gave up. Therefore I think it is important we provide a default key binding. I'm thinking ⌃⌘R due to the use of ⌘R by Spotlight.

**Comment by lhc70000** [comment]
I suggest ⇧⌘R.

We use ⇧⌘ when possible and I prefer it over ⌃⌘ because it's more comfortable when pressing. The reason we have other modifiers for some shortcuts are

- `⌘3, ⌘+, ⌘- for adjusting window size`: these are major functions and should have corresponding mpv commands, but we have custom commands because they need extra handling
- `⌃⌘P for PIP`: ⇧⌘P is occupied (show playlist)
- `⌥⌘M for music mode`: TBH I couldn't remember 😢

**Comment by low-batt** [comment]
I agree, ⇧⌘R.

**Comment by low-batt** [comment]
We also use the ⇧⌘ convention for the mpv defaults as well, but it looks like we switched to ⌥⌘ for Movist and VLC. So ⌥⌘R for Moviest and VLC?

**Comment by ShlomoCode** [comment]
> We use ⇧⌘ when possible and I prefer it over ⌃⌘ because it's more comfortable when pressing

I think it's the opposite :)

**Comment by ShlomoCode** [comment]
But Apple [says](https://developer.apple.com/design/human-interface-guidelines/keyboards#Custom-keyboard-shortcuts) "Prefer the Shift key as a secondary modifier that complements a related shortcut" and "Use the Option modifier sparingly for less-common commands or power features" 🤷

**Comment by ShlomoCode** [comment]
Note that Apple [says](https://developer.apple.com/design/human-interface-guidelines/keyboards#Custom-keyboard-shortcuts):
- Prefer the Shift key as a secondary modifier that complements a related shortcut
- Use the Option modifier sparingly for less-common commands or power features
- Avoid using the Control key as a modifier. The system uses Control in many systemwide features and shortcuts, like moving focus or capturing screenshots

**Comment by low-batt** [comment]
⇧⌘R meets those requirements, yes? We can use that for the IINA and mpv defaults. We will have to hear from others as to why the Moviest and VLC did not use that and instead used option, ⌥, instead of shift, ⇧. To match up with the existing conventions in those key bindings we should use ⌥⌘R in them. See what the other developers say.

**Comment by svobs** [comment]
> * Prefer the Shift key as a secondary modifier that complements a related shortcut
> * Use the Option modifier sparingly for less-common commands or power features

That seems almost backwards, no? Some menu items can have "alternate" versions which offer slightly different functionality. These can display an alternate title by holding down the Option key when clicking on the menu. Their key equivalent also incorporates Option.

Example in IINA: `Open In New Window…` with alternate, `Open…`

**Comment by ShlomoCode** [comment]
I added ⇧⌘R to `iina-default-input.conf`. Waiting for a decision about VLC and Movist files.

**Comment by low-batt** [comment]
Please also add ⇧⌘R to `input.conf` which is the mpv bindings.

Filtering the VLC key bindings shows for some reason ⌥⌘ was used instead of the ⇧⌘ convention:
VLC

I'd go ahead and change the Moviest and VLC key bindings to use ⌥⌘R to match up with the existing convention in those key bindings.

**Comment by ShlomoCode** [comment]
I have added

**Comment by low-batt** [comment]
I verified the key bindings. I would remove the ⇧ from the Moviest and VLC key bindings and just use ⌥⌘R to match up with many of the other key bindings that bind to IINA functions:
VLC-key

**Comment by ShlomoCode** [comment]
> I would remove the ⇧ from the Moviest and VLC key bindings and just use ⌥⌘R to match up with many of the other key bindings that bind to IINA functions

Isn't that what I did? I'm confused
>`#@iina Alt+Meta+R show-current-file-in-finder`

**Comment by ShlomoCode** [comment]
i got it. The `R` should be a lowercase letter, otherwise `Shift` is added to the shortcut. pushes a fix...

**Comment by low-batt** [comment]
Great! I will see if I can get the other developers to get this one approved and merged.

**Comment by low-batt** [comment]
@ShlomoCode Thanks for addressing #5001

**Comment by cmbail** [comment]
It doesn't appear in my file menu and doesn't work when added as a key binding. 1.3.5 Build 141|

Untitled

**Comment by low-batt** [comment]
@cmbail This change has been merged into the development version of IINA. It is not in the current release of IINA. 

Issue #5001 was closed automatically by GitHub when this pull request was merged. The downside of that is that it makes it look to users like the issue has been addressed and should be working in the latest release of IINA. I have been trying to re-open issues and leave them open until the change is in the released version of IINA, but it is exhausting to keep up with GitHub's automatic actions. GitHub users have been asking for a way to turn off this auto close feature, but last I looked GitHub has not provided a way to do that.

It is a philosophical different. GitHub views issues as tasks for developers to do. Once a change has been merged then the developer has nothing more to do and GitHub closes the issue. The IINA project views issues as the public defect and feature request database. With that point of view you would not close issues until the features are available to users. Currently the GitHub view dominates.

Sorry about the confusion.

**Comment by cmbail** [comment]
Thanks for clearing that up.

**Pull request merged**
